File: src/rngGame/main/Direction.java
package rngGame.main;

import java.util.ArrayList;
import java.util.List;

public enum Direction {

	N(-1.9635, -1.1781),
	NE(-1.1781, -0.3927),
	E(-0.3927, 0.3927),
	SE(0.3927, 1.1781),
	S(1.1781, 1.9635),
	SW(1.9635, 2.7489),
	W(2.7489, -2.7489),
	NW(-2.7489, -1.9635);

	private final double minAngle;
	private final double maxAngle;
	private boolean active = false;

	Direction(double minAngle, double maxAngle) {
		this.minAngle = minAngle;
		this.maxAngle = maxAngle;
	}

	public double getMinAngle() {
		return minAngle;
	}

	public double getMaxAngle() {
		return maxAngle;
	}

	public boolean isActive() {
		return active;
	}

	public void setActive(boolean active) {
		for (Direction direction: Direction.values()) {
			direction.active = false;
		}
		this.active = active;
	}

	public static Direction getDirectionFromAngle(double angle) {
		List<Direction> directions = new ArrayList<>();
		for (Direction direction: Direction.values()) {
			boolean isBetween = angle > direction.minAngle && angle < direction.maxAngle;
			if (
					(direction.minAngle > direction.maxAngle && (angle > direction.minAngle || angle < direction.maxAngle))
					    || isBetween
			) {
				directions.add(direction);
			}
		}
		if (!directions.isEmpty()) {
			return directions.get(0);
		} else {
			return E;
		}
	}
}

File: src/rngGame/main/Direction.kt
package rngGame.main

private const val VARIABLE_THETA_ANGLE: Double = 0.2
private const val VARIABLE_THETA_NEXT_ANGLE: Double = 0.3927

//0.3927 1.1781 1.9635 2.7489
enum class Direction(val minAngle: Double, val maxAngle: Double) {
	N(-1.9635, -1.1781),
	NE(-1.1781, -0.3927),
	E(-0.3927, 0.3927),
	SE(0.3927, 1.1781),
	S(1.1781, 1.9635),
	SW(1.9635, 2.7489),
	W(2.7489, -2.7489),
	NW(-2.7489, -1.9635);

	private var _active: Boolean = false

	var active
		get() = _active
		set(value) {
			entries.forEach {
				it._active = false
			}
			_active = value
		}

	companion object {
		fun getDirectionFromAngle(angle: Double) = (entries.firstOrNull {
				if (it.minAngle > it.maxAngle)
					angle > it.minAngle || angle < it.maxAngle
				else
					angle > it.minAngle && angle < it.maxAngle
			} ?: E).let {
				if (it.active && (angle > it.maxAngle - VARIABLE_THETA_ANGLE || angle < it.minAngle + VARIABLE_THETA_ANGLE)) {
					// Variable Theta logic
					val newAngle = if (angle > it.maxAngle - VARIABLE_THETA_ANGLE) angle + VARIABLE_THETA_ANGLE else angle - VARIABLE_THETA_ANGLE
					(entries.firstOrNull {
						if (it.minAngle > it.maxAngle)
							newAngle > it.minAngle || newAngle < it.maxAngle
						else
							newAngle > it.minAngle && newAngle < it.maxAngle
					} ?: E)
				} else {
					if (angle < it.maxAngle - VARIABLE_THETA_ANGLE && angle > it.minAngle + VARIABLE_THETA_ANGLE) {
						it.active = true
						println(it.name)
					}
					it
				}
		}
	}
}